Add check for valid freegroup link in dartsim

iche033 opened the pull request:
# 🦟 Bug fix

Fixes https://github.com/gazebosim/gz-sim/issues/2981

## Summary

Adds a check to make sure the freegroup link is valid before returning it. This fixes a crash when requesting the freegroup's root link when a nested model tree does not have any links. 

See https://github.com/gazebosim/gz-sim/issues/2981 for instructions on how to reproduce the crash.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [ ] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.)

Generated-by: Remove this if GenAI was not used.

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` and `Generated-by` messages.

azeey commented:
@osrf-jenkins run tests please

azeey reviewed (commented):
This makes me wonder why https://github.com/gazebosim/gz-sim/blob/a6b8f05ee20f75b4c65d31bb4f753bdea710c480/src/systems/physics/Physics.cc#L2469 returns a valid FreeGroup if there's no root link. i.e, if there is a valid freegroup, it should have a valid root link.

I think `FreeGroupFeatures::FindFreeGroupForModel` is probably where we should fix some logic.

iche033 commented:
> This makes me wonder why https://github.com/gazebosim/gz-sim/blob/a6b8f05ee20f75b4c65d31bb4f753bdea710c480/src/systems/physics/Physics.cc#L2469 returns a valid FreeGroup if there's no root link. i.e, if there is a valid freegroup, it should have a valid root link.
> 
> I think `FreeGroupFeatures::FindFreeGroupForModel` is probably where we should fix some logic.

I updated the logic in dartsim's `FindFreeGroupForModel` function for searching freegroups in nested models, and added test. [088990e](https://github.com/gazebosim/gz-physics/pull/765/commits/088990e721b53e7a32932eeb8461177c794cf6ae).

azeey commented on test/common_test/free_joint_features.cc:
```diff
@@ -240,6 +240,61 @@ TEST_F(FreeGroupFeaturesTest, NestedFreeGroupSetWorldPose)
+    <sdf version="1.11">
+      <model name="box">
+        <pose>1 2 3.0 0 0 0</pose>
+        <model name="nested_box"/>
```
This is not valid SDF because a non-static model has to have at least one link. Recommend adding `<static>true</static>` to the nested model.

azeey commented on test/common_test/free_joint_features.cc:
```diff
@@ -240,6 +240,61 @@ TEST_F(FreeGroupFeaturesTest, NestedFreeGroupSetWorldPose)
+    // Create the model with no links
+    sdf::ParserConfig parserConfig;
+    parserConfig.SetWarningsPolicy(sdf::EnforcementPolicy::WARN);
+    errors = root.LoadSdfString(modelStr, parserConfig);
+    EXPECT_FALSE(errors.empty()) << errors;
```
I'm not sure if reusing an `sdf::Root` object is a good idea. Probably okay, but I'd feel more comfortable with a new one.

iche033 replied:
changed to use a new `sdf::Root` object. [5c03da5](https://github.com/gazebosim/gz-physics/pull/765/commits/5c03da580a2dc4feccbb62bf573f950c11908653)

azeey reviewed (approved):
Looks good. Just a minor comment

Also running colcon CI to test this with gz-sim:
[![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ci__colcon_gpu_any-manual_ubuntu_noble_amd64&build=11)](https://build.osrfoundation.org/job/ci__colcon_gpu_any-manual_ubuntu_noble_amd64/11/)
